### ISSUE TITLE
Use code2seq to complete expression generation

### ISSUE THREAD

**Issue by CplandS** [tech-srl/slm-code-generation#2]
Hello Uri Alon，
After reading your paper and seeing that you use code2seq to achieve expression generation, there are some questions that need your help:

0. How to express the expression as a camel case representation similar to that used by code2seq.
1. How to modify the code of code2seq to make it possible to generate top5 candidates (acc5 mentioned in your paper)
Thank you.

**Comment by urialon** [comment]
Hi @CplandS ,
Thank you for your interest in our paper!

0. We split identifier and constant names by camel case, exactly as in code2seq. We train the model and predict as if each subtoken is the child of its previous subtoken. 
See the paragraph called "Generating Trees" in page3 of [the paper](https://arxiv.org/pdf/1910.00577.pdf).
In the demo website, it is presented as if all subtokens are the child nodes of the "Name" parent node:
![image](https://user-images.githubusercontent.com/15002544/87628898-9b8d9800-c73a-11ea-85c4-839bcc893aef.png)

1. We perform beam search to generate the top-k completions (see, for example, [this video](https://www.youtube.com/watch?v=UXW6Cs82UKo) for an explanation about beam search)

I hope it helps.
Uri

1.

**Comment by ShangwenWang** [comment]
Dear authors @urialon ,
Congratulations on your great work. I also have some problems about how to use this tool.:
1. How to call this tool to generate missing code? I find the _Readme_ file introduces how to train the model but I'd like to know how to execute the trained model. Is there any API or command that I can use?
2. Based on the question from @CplandS , what can I do if I want to get all the possible completions? By setting _k_ to a fixed value?

Many thanks

**Comment by urialon** [comment]
Hi @ShangwenWang , 
Thank you for your interest in our work!

1. So far, we released only the data generation and preprocessing pipeline. 
We haven't released the code that trains and executes a model yet. We hope to release that soon.
However, you **can** query the API that our [demo](https://anycodegen.org/) uses!
Try running this `curl` command:
```
curl -X POST https://w0w3uc4a63.execute-api.us-east-1.amazonaws.com/prod/predict -d '{"code": "public static Path[] stat2Paths(FileStatus[] stats) {  if (stats == null) return null;  Path[] ret = new Path[stats.length]; for (int i = 0; i < stats.length; ++i) { ret[i] = ??; } return ret; }"}'
```

To change the query code, change the code under the `"code"` property of the JSON (in my example, starting with `public static ...`). Mark the missing code with two question marks: `??`.
This will return a JSON with the 5 top completions. Copy the returned JSON into a JSON viewer [such as this one](http://jsonviewer.stack.hu/) and see its structure; I think that it is self-explanatory, but let me know if anything is unclear.

2. Do you mean in [this work](https://anycodegen.org/) or in [code2seq](https://code2seq.org/)?

**Comment by urialon** [comment]
I added this to the README: https://github.com/tech-srl/slm-code-generation/blob/master/README.md#querying-the-trained-model
Thanks for the question, this is a good idea.

**Comment by ShangwenWang** [comment]
@urialon Thanks. My second question is still about this work. 
You have introduced that the returned JSON file contains 5 top completions. I wonder what can I do if I'd like to get all the possible completions.
Thanks

**Comment by urialon** [comment]
Well, the number of possible completions is infinite. What do you mean by "all possible completions"?
I could add the beam size (=5) as a parameter for the API, but still, very large values will take a lot of time to process and the request will fail.

**Comment by ShangwenWang** [comment]
I have tested some cases I meet in my research. Some of them can be ranked at top-1 while some of them are not in the current beam size. Thus, I guess we can extend this parameter to 20 or 30. Are they "very large"?

**Comment by urialon** [comment]
Hmmm. There are two technical problems here.
Because of TensorFlow's static graph, the same graph serves all API requests. Thus, I can't have a different top-k operator for different values of K.  
Additionally, currently, the backend server is an AWS lambda which has a timeout for every request. So, larger values of beam-size will result in longer requests which might be killed by the Lambda executor. 

We will soon release the code and running inference locally will be simpler for you (because there will be no server timeout, and all queries will have the same beam size).

**Comment by ShangwenWang** [comment]
OK. Thanks for your great effort to release the code. Wish to use this tool soon.

**Comment by urialon** [comment]
Closing due to inactivity, feel free to re-open if you have any further questions!

**Comment by CplandS** [comment]
@urialon 
Hello,Uri Alon,thank you for your excellent work.
This comment is to ask when you plan to release the remaining code.

**Comment by urialon** [comment]
Hi @CplandS ,
We are working on this. 
Hopefully in a few months.
I am writing to myself to update also on this thread.

Best,
Uri

**Comment by CplandS** [comment]
Hi @urialon ,
This comment is for two questions.
1. How do you check the grammar of the generated expressions in your experiment?
2. In what month is this project scheduled to be submitted?

Thank you for your excellent job.

**Comment by CplandS** [comment]
Hi @urialon ,
This comment is for two questions.
1. How do you check the grammar of the generated expressions in your experiment?
2. In what month is this project scheduled to be submitted?

Thank you for your excellent job.

**Comment by urialon** [comment]
Hi @CplandS ,

1. We actually don't check the grammar! Initially, I implemented it such that only "valid" parent-child are allowed. For example, if the current expanded node is of type `A` and `A` can have only `{B,C}` as valid child nodes, when predicting the child of `A` I performed the softmax only over `B` and `C`. Later, I found that the model performs slightly better if I don't restrict it to `B` and `C` only, and let the model choose any node type. In practice, syntactic errors were rare.

2. We don't have an estimate yet, sorry. 

Best,
Uri
